x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/metadata/MetadataMigrateToDataTiersRoutingService.java

```diff
@@ -0,0 +1,280 @@
+                                phase.getName(), allocateAction);
+                        } else {
+                            // remove the action altogether
+                            actionMap.remove(allocateAction.getWriteableName());
+                            logger.debug("ILM policy [{}], phase [{}]: removed the allocate action", lifecyclePolicy.getName(),
+                                phase.getName());
+                        }
+
+                        Phase updatedPhase = new Phase(phase.getName(), phase.getMinimumAge(), actionMap);
+                        Map<String, Phase> updatedPhases =
+                            new HashMap<>(newLifecylePolicy == null ? lifecyclePolicy.getPhases() : newLifecylePolicy.getPhases());
+                        updatedPhases.put(phaseEntry.getKey(), updatedPhase);
+                        newLifecylePolicy = new LifecyclePolicy(lifecyclePolicy.getName(), updatedPhases);
+                    }
+                }
+
+                if (newLifecylePolicy != null) {
+                    // we updated at least one phase
+                    long nextVersion = policyMetadataEntry.getValue().getVersion() + 1L;
+                    LifecyclePolicyMetadata newPolicyMetadata = new LifecyclePolicyMetadata(newLifecylePolicy,
+                        policyMetadataEntry.getValue().getHeaders(), nextVersion, Instant.now().toEpochMilli());
+                    LifecyclePolicyMetadata oldPolicyMetadata = newPolicies.put(policyMetadataEntry.getKey(), newPolicyMetadata);
+                    assert oldPolicyMetadata != null :
+                        "we must only update policies, not create new ones, but " + policyMetadataEntry.getKey() + " didn't exist";
+
+                    updateIndicesForPolicy(mb, currentState, xContentRegistry, client, oldPolicyMetadata.getPolicy(), newPolicyMetadata);
+                    migratedPolicies.add(policyMetadataEntry.getKey());
+                }
+            }
+
+            IndexLifecycleMetadata newMetadata = new IndexLifecycleMetadata(newPolicies, currentLifecycleMetadata.getOperationMode());
+            mb.putCustom(IndexLifecycleMetadata.TYPE, newMetadata);
+        }
+        return migratedPolicies;
+    }
+
+    static boolean allocateActionDefinesRoutingRules(String nodeAttrName, @Nullable AllocateAction allocateAction) {
+        return allocateAction != null && (allocateAction.getRequire().get(nodeAttrName) != null ||
+            allocateAction.getInclude().get(nodeAttrName) != null ||
+            allocateAction.getExclude().get(nodeAttrName) != null);
+    }
+
+    static List<String> migrateIndices(Metadata.Builder mb, ClusterState currentState, String nodeAttrName) {
+        List<String> migratedIndices = new ArrayList<>();
+        String nodeAttrIndexRoutingSetting = INDEX_ROUTING_REQUIRE_GROUP_SETTING.getKey() + nodeAttrName;
```

Comment:
For indices I'm only looking at the `require.data` routing setting (as opposed to also looking at `include.data` and `exclude.data`) even though for ILM policies _any_ allocate action routing configuration for the `data` attribute will trigger the migration of the ILM policy.

The reasoning behind this is that routing configurations for indices can come from multiple places (which is part of the reason why we need to migrate at all :) ) as opposed to ILM policies which are meant to move forward an existing routing configuration (eg. the mantra in ILM might be more like "take it from the warm phase onwards and own it"). 

Even more than that, it'd be extremely difficult to know what the equivalent `_tier_preference` configuration would be for an index that has `exclude.data:cold,warm,custom` (and all the other myriad of options that might be out there)

@dakrone do you think it makes sense like this or should we try to do more for indices? 

I was pondering maybe also reporting the indices that we did NOT migrate. What do you think?

Reply:
>  is there a reason we would want to know this information?

I was thinking that with many indices in a deployment, _some_ might not adhere to the rules and structure we expect. So it might be difficult to spot the 3 indices out of 300 that we couldn't migrate - until they can't be allocated. 

We currently issue a `warn` logging so maybe that's enough for now (?)

Reply:
> I think we should inspect require and include (skipping exclude)